### PR TITLE
build: add tsc type checks and emit declarations in the ci/test pipeline

### PR DESCRIPTION
**Pull request by lance**
This commit introduces TypeScript checks and generates type declarations for the existing JavaScript codebase using `tsc` prior to running the linter task. This is similar to [@grant's pull request](https://github.com/cloudevents/sdk-javascript/pull/137), but keeps the code base as pure JS mainly by just keeping the entry point `index.js` instead of `.ts`.

I am proposing this as a very slight alternative. Since I'm not fully familiar with the TypeScript tooling, I wasn't sure if the type declaration files should live side-by-side with the JS or not, but I assume they should be kept in SCM. Feedback welcome.

Ref: https://github.com/cloudevents/sdk-javascript/issues/9

### CODE REVIEW

**Comment by grant** [comment]
If we're going the "never change `.js` to `.ts` at all costs" route, I would at least do the following:
- Git ignore all generated files. These `.d.ts` files are not hand-written, they are generated.

I'm surprised that you're ok with generated `.ts` file bloat, but not normal `.ts` files that would do the exact same thing. If you just renamed the files, you wouldn't see duplicated files and large PRs...

**Comment by lance** [comment]
@grant again I think you are not representing my position in good faith. I have never said that we are going the "never change `.js` to `.ts` at all costs" route. That said, thanks for your feedback. 

I humbly asked what was the right thing to do with those `.d.ts` files, so there's no need to be snarky about bloat and again dismissive of my other concerns such as existing maintainer alienation and the potential pool.

**Comment by grant** [comment]
@lance Oh, I'm not trying to be snarky. I'm trying to get TypeScript support here.

It's common to not include generated code in git repositories, as it can simply be generated from the source code again. In this case, we are generating files with the extension `.d.ts`. These files should not be included with the PR.

Emitting declarations is fine for local development but shouldn't be seen on GitHub or npm.

Again, this is a weird way to beat around the bush and legacy TypeScript knowledge. Ideally, like most modules, we don't need any `.d.ts` files.

See the handbook for more details:
https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,69 @@
+{
```
Comment (grant) [inline tsconfig.json:1]: It's usually common to delete the comments after a `tsc --init`. These are just helpers for configuring your setup.

File: package.json [hunk 2]
```diff
@@ -108,7 +111,8 @@
     "nock": "~12.0.3",
     "nyc": "~15.0.0",
     "standard-version": "^7.1.0",
-    "standardx": "^5.0.0"
+    "standardx": "^5.0.0",
+    "typescript": "^3.8.3"
```
Comment (grant) [inline package.json:115]: `typescript` should be a dev dependency, not a dependency.

File: package.json [hunk 3]
```diff
@@ -93,6 +95,7 @@
   },
   "homepage": "https://github.com/cloudevents/sdk-javascript#readme",
   "dependencies": {
+    "@types/node": "^13.13.5",
```
Comment (grant) [inline package.json:98]: All `@types` modules should not be included in a dependency. It has no runtime functionality.

**Comment by lance** [comment]
> It's common to not include generated code in git repositories, as it can simply be generated from the source code again. In this case, we are generating files with the extension .d.ts. These files should not be included with the PR.

OK - I will update it.

> Emitting declarations is fine for local development but shouldn't be seen on GitHub or npm.

The handbook you link to is all about publishing your `.d.ts` files to npm, so I'm confused by this statement.

File: package.json (hunk 2 above)
Comment (lance) [reply package.json:115]: That's where it is already, in the `devDependencies` section?

File: package.json (hunk 3 above)
Comment (lance) [reply package.json:98]: I think this should be a `devDependency` since `tsc` needs it when type checking.

**Comment by grant** [comment]
> The handbook you link to is all about publishing your `.d.ts` files to npm, so I'm confused by this statement.

Yes. The link has really useful guides in using TypeScript, including best practices. The issue about files on npm is just one of the comments on the added `dependencies` in this PR. I recommend reading it.

It will clearly recommend using TypeScript `.ts` files for dozens of benefits.

**Comment by grant** [comment]
Are we planning on ignoring https://github.com/cloudevents/sdk-javascript/pull/137/?
This Pull Request, https://github.com/cloudevents/sdk-javascript/pull/155 will clearly conflict with that one.

I don't understand.

File: package.json (hunk 3 above)
Comment (grant) [reply package.json]: Yes, it's always a dev dep.

**Comment by lance** [comment]
> I don't understand.

As I wrote in the description :man_shrugging: 

> I am proposing this as a very slight alternative.

File: package.json (hunk 2 above)
Comment (grant) [reply package.json:115]: Oops. Looks right then.
The GitHub tool collapsed the file making it look like a dep.

**Comment by grant** [comment]
We can't just ignore people's pull requests and create/merge your own without discussion!

Why can't we just pool together experience and collaborate!

**Comment by lance** [comment]
@grant I'm not ignoring your PR - I'm proposing an alternative because we have been doing nothing but discussing this for a while now. There has been a lot of discussion. I wish you could acknowledge that I'm at least trying to find some common ground here.

**Comment by grant** [comment]
Thanks Lance for listening through these discussions.

Can you review the PR and not create new PRs that conflict with old ones?

**Comment by erikerikson** [comment]
Thank you @lance for providing this as an alternative and clearly marking it as such.

@grant it's not clear where Lance has merged his own PRs (although pretty obvious he created this one).  As he's indicated the point of this is to give substance to the discussion that has been ongoing and he does seem to be acting in good faith.

**Comment by cwallsfdc** [comment]
@lance, I appreciate that you are compromising.  But, if you are willing to run `tsc` and are now familiar w/ `tsconfig.json` and also understand the value of typings, I think we should favor @grant's PR.

Can we move to `.ts` and relax typical Typescript requirements in `tsconfig.json` so that those who contribute as JS authors still feel empowered to do so?  Is it possible to be `.ts` and still follow JS constructs?  If so, I'm confident that JS developers can still be productive in `.ts` files.  And, my hope is that, over time, they'll appreciate the value-add that Typescript bring.

**Comment by lance** [comment]
@cwallsfdc can you help me understand what doing that actually achieves if the stated intent is to stick with pure JS syntax? Why go through the transpilation step if we're going from pure JS in a `.ts` file to pure JS in a `.js` file? Does the enabling of code completion through type definitions and quality JSdoc not get us to a user experience like the one that Grant has been looking for?

I realize that you and others may find it hard to understand why I don't want to _immediately_ move to `.ts` and code transpilation, although I have stated that I see this as a first step on a possible continuum. But similarly, I find it hard to understand what that next step gets us. What is gained? 

Isn't the goal to have a better experience for users of this SDK? If so, how does my alternative approach not achieve that in an equivalent way while also allowing room to move forward with something further along the continuum later? I'm just asking for some compromise. I've conceded a lot since the beginning of this discussion. I would like to think that the community also gives credence to my concerns.

**Comment by lance** [comment]
> help me understand what doing that actually achieves if the stated intent is to stick with pure JS syntax?

It seems that this is a fundamental misunderstanding on my part. It's become quite clear that this is not the actual intent.

**Comment by lance** [comment]
code-completion
Just so it's tangible, here is code completion in the IDE.

**Comment by erikerikson** [comment]
Including the JsDocs for immediately answering semantic questions non-experts and those trying out the SDK may have.  Regardless of the final approach to tooling, I expect that enriched JsDocs will be a positive addition to support adoption and ongoing usage.

**Comment by grant** [comment]
> > help me understand what doing that actually achieves if the stated intent is to stick with pure JS syntax?
> 
> It seems that this is a fundamental misunderstanding on my part. It's become quite clear that this is not the actual intent.

It is the intent. What do you mean? You are insinuating malintent.

I can imagine we could add a TS interface for a CloudEvent object in the future if agreed upon. How else will we see object autocompletion without getters and setters?

**Comment by lance** [comment]
@grant I honestly do appreciate you making this compromise, helping us both find a path forward towards making this project what we both want it to be -- a high quality, stable and and user-friendly module. Though our conversation has been heated at times, I look forward to putting all of that behind us and working together with you and the rest of the community to make that happen.

**Pull request merged**